tests/suites/os/tests/secureboot/index.js

```diff
@@ -1,145 +1,193 @@
-						 *
-						 * Otherwise, the output finishes the above background color escape
-						 * sequence, clears the screen, and shows 'loading Boot0003
-						 * "balenaOS"'
-						 */
-
-							waitForSerialOutput(/\x1B\[0m\x1B\[37m\x1B\[40m$/, 3), // eslint-disable-line no-control-regex
-							'Bootloader will not load configuration that fails signature verification',
-						{ todo: 'This needs reworked for GRUB 2.12' },
-						)
-					).then(resetWorker);
-				};
-
-				if (securebootSupported) {
-					if (securebootEnabled) {
-						await testFullDiskEncryption();
-						await testModuleVerification();
-						await testBootloaderIntegrity();
-						await testBootloaderConfigIntegrity();
-					} else {
-						test.comment('Secure boot is not enabled');
-					}
-				} else {
-					test.comment('Secure boot is not supported by firmware');
-				}
+						/crypt/,
+						`${args.name} is encrypted`,
+					)
+				)
+			);
+	}
+
+	async testModuleVerification() {
+		await this.test.resolves(
+			this.worker.executeCommandInHostOS('modprobe pcan_netdev', this.link),
+			'Module with valid signature loads',
+		);
+
+		await this.worker.pushContainerToDUT(
+			this.link,
+			`${__dirname}/kernel-module-build`,
+			'load',
+		);
+
+		return this.test.resolveMatch(
+			this.worker.executeCommandInHostOS(
+				'balena logs $(balena ps -aqf NAME=load)',
+				this.link,
+			),
+			/Key was rejected by service/,
+			'Unsigned module does not load',
+		);
+	}
+
+	async testBootloaderIntegrity() {
+		throw new Error("Method testBootloaderIntegrity() must be implemented");
+	}
+
+	async testBootloaderConfigIntegrity() {
+		throw new Error("Method testBootloaderConfigIntegrity() must be implemented");
+	}
+}
+
+class uefiSecureBoot extends secureBoot {
+	async isSecureBootSupported() {
+			return this.worker.executeCommandInHostOS(
+				['ls', securebootEfiVarPath, '&>/dev/null',
+					'&&', 'echo', 'true',
+					'||', 'echo', 'false'
+				],
+				this.link,
+			).then(output => { return output === 'true'});
+	}
+
+	async isSecureBootEnabled() {
+			// https://wiki.archlinux.org/title/Unified_Extensible_Firmware_Interface/Secure_Boot#Checking_Secure_Boot_status
+			return this.worker.executeCommandInHostOS(
+					[
+						'od', '--address-radix=n', '--format=u1', securebootEfiVarPath,
+							'|', 'tr', '-s', "' '",
+							'|', 'cut', "-d' '", '-f6'
+					],
+					this.link,
+			).then(output => output === '1' );
+	}
+
+	async testBootloaderIntegrity() {
+		return this.worker.executeCommandInHostOS(
+			['balena', 'run', '--rm', '-v', '/mnt:/mnt', 'alpine', '/bin/sh', '-c',
+				'"apk add --update --no-cache binutils && strip /mnt/boot/EFI/BOOT/bootx64.efi"'],
+			this.link,
+		).then(() => this.worker.executeCommandInHostOS('reboot', this.link)
+		).then(() => this.test.resolves(
+			this.waitForSerialOutput(/bootx64.efi: Access Denied/),
+			'Firmware will not load bootloader that fails verification',
+		)
+		).then(this.resetWorker());
```

Comment:
@alexgg This is broken. `.then()` accepts a callback function for when the promise completes. The original code is written as `.then(resetWorker)`, which means "execute resetWorker and return the resulting promise". This code executes the function immediately, and the resulting promise is not a callable handler for the previous promise resolving.

Reply:
I suspect the effect this has is resetting the DUT as soon as the statement is evaluated, rather than after the test is completed.

Reply:
indeed @jakogut thanks - I changed it to using an arrow function